docs/proposals/object-store.md

```diff
@@ -0,0 +1,161 @@
+Adding Capability for BYO Object Store
+================
+
+This document outlines the current architecture of how Object Stores are used by [kots](https://github.com/replicatedhq/kots) and [kotsadm](https://github.com/replicatedhq/kotsadm), and a proposal for modifying how it is configured to allow end users of `kotsadm` to bring their own object store rather than using an embedded minio instance.
+
+### Goals
+
+- For `kots install` and `kots pull`, allow an end user to bring their own object store rather than using an embedded minio instance that requires the cluster to include a storage class.
+- Move closer to unifying the YAML used across `kots install` and the [kURL addon](https://kurl.sh/docs/add-ons/kotsadm) (or, at the very least, avoid creating any more differences between the two methods)
+
+### Architecture Background
+
+An Object store is required by kots to store each version of the deployable application. For each version deployed, a tar bundle is pushed to the object store which contains the `upstream`, `base`, and `overlays` directories from the application. This includes any end-user specified `Config`, `Application`, and/or `License` state stored in `upstream/userdata`, as well as any custom last-mile [kustomize](https://kustomize.io) changes made in `overlays/`.
+
+
+When a new upstream version is detected by kotsadm, either via polling or via an airgap bundle upload, the new upstream version will be templated out, with the user-supplied config and overlays from the previous version being applied to the new version from upstream. Once the upstream has been inflated to `upstream`, `base`, and `overlays`, it is bundled and pushed to a new key in the object store.
+
+Versions stored in the object store are used
+
+1. To deploy a specific version of the application after the upstream has been processed
+1. To roll back to previous versions of the application (if [allowRollback](https://kots.io/reference/v1beta1/application/#allowrollback) is specified).
+
+
+
+Current Architecture 
+---------------
+
+In every case, five environment variables are configured for `kotsadm` and `kotsadm-api`:
+
+- `S3_ACCESS_KEY_ID`
+- `S3_SECRET_ACCESS_KEY`
+- `S3_ENDPOINT`
+- `S3_BUCKET_NAME`
+- `S3_BUCKET_ENDPOINT` 
+
+`S3_BUCKET_ENDPOINT` appears to be a boolean, configured whether to use a custom endpoint. Seems to always be `true`, I believe this controls whether to override the default S3 endpoint in our S3 client.
```

Comment:
My understanding of this param is that it controls whether the client (our s3 internal client) uses path style or dns style bucket naming.  (https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story)

I believe that minio doesn't currently support bucket dns names in K8s, so we rely on this legacy pattern of addressing buckets

Reply:
ahhhh okay gotcha -- do you think we need an extra flag for this then?